### ISSUE TITLE
Forecasting the unknown future

### ISSUE THREAD

**Issue by kthouz** [statsmodels/statsmodels#3212]
Hi, 
First of all thanks for putting this ARIMA summary online.
Second, I have a question. How would someone make predictions on the future (test set) if observed data are still unknown (NaN). Here is a sample code I tried https://github.com/kthouz/Misc/blob/master/arima_tuto.ipynb

**Comment by ChadFulton** [comment]
I'm not quite clear what you're trying to do, but maybe you're looking for the `forecast` method? i.e. you want to predict the future from a known dataset? In that case, something like;

```python
mod = sm.tsa.statespace.SARIMAX(endog, exog=exog, order=(1,0,1))
res = mod.fit()
forecasts = res.forecast(10)
```

**Comment by kthouz** [comment]
Briefly, given historical data from 1990 to 1999, I want to make predictions on year 2000. Here is the code I tried

`# get the data
data = pd.read_csv('https://raw.githubusercontent.com/kthouz/Misc/master/15.csv')
data.index = pd.DatetimeIndex(data.Date)
data.drop('Date',axis=1,inplace=True)
#
# split into train, validation and test
train = data.loc[:'1999-12-31']
valid = data.loc['2000-01-01':'2001-12-31']
test = data.loc['2002-01-01':] # test set has just NaNs
#
# train model
mod = sm.tsa.statespace.SARIMAX(endog=train.Rides,order=(1,0,1))
res = mod.fit()
#
# make predictions
predictions = res.predict(start = '1999-01-01', end = '2000-12-31')
plt.plot(train.loc['1997-01-01':,'Rides'],'.') #plot data of the last three years
plt.plot(predictions) #overlay the predictions from one last year up to 1 year in the future
plt.xticks(rotation='vertical')
plt.legend(['data','predictions'])
plt.show()`

And the output looks strange. Have a look at January 2000 to December 2000
![demo](https://cloud.githubusercontent.com/assets/9649515/18875853/7aab9588-8495-11e6-9c67-d4250357a86a.png)

**Comment by ChadFulton** [comment]
That seems reasonable to me, the model you specified has no seasonal component, so the forecast will simply revert back to the mean, which it appears to be doing. The SARIMAX class does allow the introduction of a seasonal component. Also, I think you will want to add a constant term to your model or else de-mean your data, since the mean is definitely not zero.

**Comment by kthouz** [comment]
Thanks a lot. I made the correction and it works pretty okay. However, I had to use monthly averages because, the model was taking forever to converge for a seasonality of 365 days.
Here is the line i changed
'mod = sm.tsa.statespace.SARIMAX(endog=train.Rides,trend='n',order=(1,0,1),seasonal_order=(1,0,1,12))'

to get
![demo](https://cloud.githubusercontent.com/assets/9649515/18899033/ead74d84-8503-11e6-8bce-932c95e497c3.png)

**Comment by ChadFulton** [comment]
Yes, due to the state space setup large seasonal periods can take a long time. You could use the `simple_differencing=True` argument, although then you will have 365 fewer datapoints on which to estimate the parameters.

**Comment by sherrylau** [comment]
Hi,

I am also trying to forecast future values by using this:
```{python}
mod = sm.tsa.statespace.SARIMAX(endog, exog=exog, order=(1,0,1))
res = mod.fit()
forecasts = res.forecast(10)
```
however i got an error as follow:

  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/base/wrapper.py", line 95, in wrapper
    obj = data.wrap_output(func(results, *args, **kwargs), how)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/statespace/mlemodel.py", line 2411, in forecast
    return self.predict(start=self.nobs, end=end, **kwargs)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/statespace/mlemodel.py", line 2384, in predict
    prediction_results = self.get_prediction(start, end, dynamic, **kwargs)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/statespace/sarimax.py", line 1878, in get_prediction
    _start = self.model._get_predict_start(start)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/base/tsa_model.py", line 141, in _get_predict_start
    self._set_predict_start_date(start)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/base/tsa_model.py", line 115, in _set_predict_start_date
    1, self.data.freq)
  File "/Users/212448740/anaconda/lib/python2.7/site-packages/statsmodels-0.8.0-py2.7-macosx-10.6-x86_64.egg/statsmodels/tsa/base/datetools.py", line 83, in _date_from_idx
    return _maybe_convert_period(d1) + idx * _freq_to_pandas[freq]
TypeError: unsupported operand type(s) for *: 'int' and 'NoneType'

**Comment by ChadFulton** [comment]
This has to do with a current bug in Statsmodels if you have certain date-related indexes without a frequency. You can either re-index your data to have a defined frequency or you can use numpy arrays.

For example, the following should work:

```python
mod = sm.tsa.statespace.SARIMAX(endog.values, exog=exog.values, order=(1,0,1))
res = mod.fit()
forecasts = res.forecast(10)
```

**Comment by makmanalp** [comment]
As of recently, the import that works for me is `from statsmodels.tsa.statespace.sarimax import SARIMAX`

**Comment by makmanalp** [comment]
Also, some cool SARIMAX examples for those like me who came here looking for ARIMA forecasting: http://www.statsmodels.org/dev/examples/notebooks/generated/statespace_sarimax_stata.html
